collect_app/src/main/java/org/odk/collect/android/logic/FormController.java

```diff
@@ -683,22 +683,26 @@ public int stepToOuterScreenEvent() {
     public FailedConstraint saveAllScreenAnswers(HashMap<FormIndex, IAnswerData> answers,
                                                  boolean evaluateConstraints) throws JavaRosaException {
         if (currentPromptIsQuestion()) {
+            // save answers
             for (FormIndex index : answers.keySet()) {
                 // Within a group, you can only save for question events
                 if (getEvent(index) == FormEntryController.EVENT_QUESTION) {
-                    int saveStatus;
-                    IAnswerData answer = answers.get(index);
-                    if (evaluateConstraints) {
-                        saveStatus = answerQuestion(index, answer);
+                    saveAnswer(index, answers.get(index));
+                } else {
+                    Timber.w("Attempted to save an index referencing something other than a question: %s",
+                            index.getReference().toString());
+                }
+            }
+            if (evaluateConstraints) { // check if answers are valid
+                int saveStatus;
+                for (FormIndex index : answers.keySet()) {
+                    // Within a group, you can only save for question events
+                    if (getEvent(index) == FormEntryController.EVENT_QUESTION) {
+                        saveStatus = answerQuestion(index, answers.get(index));
```

Comment:
This code looks ok to me. I agree that it's not ideal that answerQuestion re-saves the answer. We could do with a "checkAnswer" that returns the status without the save.